### ISSUE TITLE
HKG with Haskell - rewrite lemma with function in LHS

### ISSUE THREAD

**Issue by virgil-serbanuta** [runtimeverification/evm-semantics#726]
The Haskell backend cannot unify functions, so it cannot find a value for the `MEM'` variable, so it crashes. The variable is wrapped in the `#symCmem` function.

https://github.com/kframework/evm-semantics/blob/master/tests/specs/erc20/abstract-semantics-segmented-gas.k#L26

**Comment by denis-bogdanas** [comment]
In principle `[matching]` can be replaced with unification lemmas. Like this:
```

**Comment by yzhang90** [comment]
the lemma says `#symCmem` is an injective function, which is not what we want.

**Comment by denis-bogdanas** [comment]
In VSC `#symCmem` is no longer used. Only `#gas` is used as argument to `[matching]`. But `#gas` is not injective so such lemmas cannot be used.

**Comment by yzhang90** [comment]
@ehildenb since the #gas is a wrapper for the three arguments `INITGAS, NONMEM, MEM`, and we know that the arguement in `MEM` can cancel with each other during the execution. It is simply a workaround since both the backend doesn't support AC for arithmetic. Can we mark the #gas as constructor for now?

**Comment by ehildenb** [comment]
I mean, it is a constructor isn't it? I still would like the actual semantics to be given in terms of `#gas(_,_,_)`, instead of only for proving.

I think it should be pretty straightforward to do:

```
    rule <k> #gas [ OP , AOP ] 
          => #if #usesMemory(OP) #then #memory [ AOP ] #else .K #fi 
          ~> #gas [ AOP ] 
          ~> #resolveGas
         ... 
        </k> 

    rule <k> #gas [ OP ] => #gasExec(SCHED, OP) ~> #deductGas ... </k> 
         <schedule> SCHED </schedule>  

    rule <k> #memory [ OP ] => #memory(OP, MU) ~> #deductMemory ... </k> 
         <memoryUsed> MU </memoryUsed> 

    rule <k> MU':Int ~> #deductMemory => (Cmem(SCHED, MU') -Int Cmem(SCHED, MU)) ~> #deductMemoryGas ... </k> 
         <memoryUsed> MU => MU' </memoryUsed> <schedule> SCHED </schedule> 
 
    rule <k> G:Int ~> #deductMemoryGas => .  ... </k>
          <gas> #gas(_, _, MEMGAS => MEMGAS +Int G) </gas> 
    rule <k> G:Int ~> #deductGas => . ... </k>
          <gas> #gas(_, EXECGAS => EXECGAS +Int G, _) </gas>

    rule <k> #resolveGas => . ... </k>
          <gas> #gas(G, EXECGAS, MEMGAS) => #gas(G -Int EXECGAS -Int MEMGAS) </gas>
       requires G >=Int (EXECGAS +Int MEMGAS)

    rule <k> #resolveGas => #end EVMC_OUT_OF_GAS ... </k>
         <gas> #gas(G, EXECGAS, MEMGAS) </gas>
       requires notBool (G >=Int (EXECGAS +Int MEMGAS))
```

Then for verification we can intercept and disable the `#resolveGas` rule.

**Comment by yzhang90** [comment]
ok. Here are my thoughts:

1. This change requires us to put `#gas(X, 0, 0)` in the <gas> cell for both concrete and symbolic execution. is that ok?
2. We introduce the #gas only because the backend doesn't support AC. Suppose one day the backend supports AC, do we need to change it back? Or #gas representation is clearer?
3.  When we use #gas representation during proofing, we implicitly mean that the gas is always sufficient. And as a result, there will be no OUT_OF_GAS exception during the symbolic execution. Do we want the same?
4. I think it is a good solution if the above answers are yes. And I would mark
```
    rule <k> #resolveGas => . ... </k>
         <gas> #gas(G, EXECGAS, MEMGAS) => #gas(G -Int EXECGAS -Int MEMGAS, 0, 0) </gas>
       requires G >=Int (EXECGAS +Int MEMGAS)

    rule <k> #resolveGas => #end EVMC_OUT_OF_GAS ... </k>
         <gas> #gas(G, EXECGAS, MEMGAS) </gas>
       requires notBool (G >=Int (EXECGAS +Int MEMGAS))
``` 
as concrete and put
```
    rule <k> #resolveGas => . ... </k> [trusted]
```
in the `verification.k`

**Comment by ehildenb** [comment]
Yes, for both symbolic and concrete is OK. I would rather have as small change between concrete and symbolic as possible, so if that means concrete needs to change to accomodate symbolic, let's do it.

If AC reasoning gets better, we can update the semantics to remove this distinction.

And yes, I think having the option to turn off gas verification (by adding the rule `rule <k> #resolveGas => . ... </k> [priority(25)]` is good. That way we can get proofs done first, then do gas verification afterwards once we have the accumulated gas expression.

**Comment by ehildenb** [comment]
@yzhang90 note that another way is to change the configuration:

```
configuration
    ...
    <gas>
      <gasAvail> 0:Int </gasAvail>
      <gasExec> 0:Int </gasExec>
      <gasMem> 0:Int </gasMem>
    </gas>
```

Also, instead of using `#resolveGas`, we can add:

```
syntax Gas ::= "InfGas" | Int | deductGas ( Gas , Int ) [function]
rule deductGas(G:Int, G':Int) => G -Int G'
rule deductGas(InfGas, _) => InfGas
```

This would allow the exact same way for both symbolic and concrete execution, _and_ would help with an issue we have with the Firefly client where we need to be able to execute a contract without ever running out of gas. So I think it's a good approach.

**Comment by yzhang90** [comment]
To sum up, following Everett's first approach, we modify the semantics as follows:

```
    syntax Gas := #gas(AvailGas, Int, Int)
    syntax AvailGas := "InfGas" | Int 

    rule <k> #gas [ OP , AOP ] 
          => #if #usesMemory(OP) #then #memory [ AOP ] #else .K #fi 
          ~> #gas [ AOP ] 
          ~> #resolveGas
         ... 
        </k> 

    rule <k> #gas [ OP ] => #gasExec(SCHED, OP) ~> #deductGas ... </k> 
         <schedule> SCHED </schedule>  

    rule <k> #memory [ OP ] => #memory(OP, MU) ~> #deductMemory ... </k> 
         <memoryUsed> MU </memoryUsed> 

    rule <k> MU':Int ~> #deductMemory => (Cmem(SCHED, MU') -Int Cmem(SCHED, MU)) ~> #deductMemoryGas ... </k> 
         <memoryUsed> MU => MU' </memoryUsed> <schedule> SCHED </schedule> 
 
    rule <k> G:Int ~> #deductMemoryGas => .  ... </k>
         <gas> #gas(_, _, MEMGAS => MEMGAS +Int G) </gas>

    rule <k> G:Int ~> #deductGas => . ... </k>
         <gas> #gas(_, EXECGAS => EXECGAS +Int G, _) </gas>

    rule <k> #resolveGas => . ... </k>
         <gas> #gas(G:Int , EXECGAS, MEMGAS) => #gas(G -Int EXECGAS -Int MEMGAS, 0, 0) </gas>
       requires G >=Int (EXECGAS +Int MEMGAS)

    rule <k> #resolveGas => #end EVMC_OUT_OF_GAS ... </k>
         <gas> #gas(G:Int , EXECGAS, MEMGAS) </gas>
       requires notBool (G >=Int (EXECGAS +Int MEMGAS))

    rule <k> #resolveGas => . ... </k>
         <gas> #gas(InfGas, EXECGAS, MEMGAS) </gas>
```
Note that the last rule is specially designed for cases where we don't want to run out of gas (like verification and etc.) and still want to get a precise gas expression.

Another option is to change #gas(AvailGas, Int, Int) into 
```
configuration
    ...
    <gas>
      <gasAvail> 0:Int </gasAvail>
      <gasExec> 0:Int </gasExec>
      <gasMem> 0:Int </gasMem>
    </gas>
``` and follow the similar approach, however, i think this may require change a lot of specs.

**Comment by ehildenb** [comment]
Also note that having `InfGas` is useful for the Firefly tool.

**Comment by denis-bogdanas** [comment]
Looks good. I'd go with first option to minimize changes to specs. Please don't push this change before I wrap up Solar. Will do it today.

**Comment by anvacaru** [comment]
Fixed by #735.
